File: lib/provenance.js
const request = require('request');
const fs = require('fs');
const config = require('config');
const utils = require('../lib/utils');
const logger = require('../config/winston');

class Provenance {

  static httpRequest(method, action, params, port, callback) {

    const URL = config.get('provenance_server.URL').replace("8081", port) + "/" + action;
    const headers = {'Content-Type' : 'application/json'}

    request({

      url: URL,
      headers: headers,
      method: method,
      qs: params

    }, function (error, response, body) {

      if ( error || (response && response.statusCode >= 400) ) {

        logger.error("Cannot contact provenance server. " + error + " " + ( ( body && typeof body === "object" ) ? JSON.stringify(body) : body ) + " " + ( response && response.statusCode ? response.statusCode : "Status unknown." ));
        callback({});

      } else {

        callback(body);

      }

    });

  }

  static get(action, params, port, callback) {

    this.httpRequest("GET", action, params, port, callback)

  }

  static post(action, params, port, callback) {

    this.httpRequest("POST", action, params, port, callback)

  }

  static new(identifier, default_namespace, port, callback=function(response){}) {

    const params = {'identifier' : identifier, 'defaultNamespace' : default_namespace}
    this.post('new', params, port, function(response) { callback(response) })

  }

  static namespace(identifier, prefix, namespace, port, callback=function(response){}) {

    const params = {'identifier' : identifier, 'prefix' : prefix, 'namespace' : namespace}
    this.post('namespace', params, port, function(response) { callback(response) });

  }

  static register(document_identifier, template_identifier, data, port,  callback=function(response){}) {

    const params = {'documentIdentifier' : document_identifier, 'templateIdentifier' : template_identifier, 'templateData' : data}
    this.post('register', params, port, function(response) { callback(response) })

  }

  static generate(document_identifier, template_identifier, fragment_identifier, data, port, callback=function(response){}) {

    const params = {'documentIdentifier' : document_identifier, 'templateIdentifier' : template_identifier, 'fragmentIdentifier' : fragment_identifier, 'data' : data}
    this.post('generate', params, port, function(response) { callback(response) })

  }

  static listDocuments(port, callback=function(response){}) {

    this.get('list', {}, port, function(response) { callback(response) });

  }

  static delete(identifier, port, callback=function(response){}) {

    const params = {'identifier' : identifier}
    this.post('delete', params, port, function(response) { callback(response) })

  }

  static exportItem(identifier, doctype, port, callback=function(response){}) {

    const params = {'identifier' : identifier, 'docType' : doctype}
    this.get('export', params, port, function(response) { callback(response) })

  }

  static add(id, fragment, templateID, templatePath, port, callback) {

    const documentID = "data-" + id;
    const fragmentID = "frag-" + id;

    var fragment = utils.replaceAll(fragment, "\\[id\\]", id);

    Provenance.new(documentID, 'http://name.space', port, function(newBody) {

      Provenance.namespace(documentID, 'sub', 'http://sub.name.space', port, function(namespaceBody) {

        const template = fs.readFileSync(templatePath, 'utf8');

        Provenance.register(documentID, templateID, template, port, function(registerBody) {

          Provenance.generate(documentID, templateID, fragmentID, fragment, port, function(generateBody) {

            callback(generateBody);

          });

        });

      });

    });

  }

}

module.exports = Provenance;

File: lib/provenance.py
#!/usr/bin/env python

import json
import requests

def post(method, params, port):
    url = 'http://localhost:' + str(port) + '/'
    headers = {'Content-Type' : 'application/json'}
    r = requests.post(url + method, headers=headers, params=params)
    return r.text

def get(method, params, port):
    url = 'http://localhost:' + str(port) + '/'
    headers = {'Content-Type' : 'application/json'}
    r = requests.get(url + method, headers=headers, params=params)
    return r.text

def new(identifier, default_namespace, port):
    params = {'identifier' : identifier, 'defaultNamespace' : default_namespace}
    return post('new', params, port)

def namespace(identifier, prefix, namespace, port):
    params = {'identifier' : identifier, 'prefix' : prefix, 'namespace' : namespace}
    return post('namespace', params, port)

def register(document_identifier, template_identifier, data, port):
    params = {'documentIdentifier' : document_identifier,
            'templateIdentifier' : template_identifier, 'templateData' : data}
    return post('register', params, port)

def register_template(document_identifier, template_identifier, port):
    params = {'documentIdentifier' : document_identifier,
            'templateIdentifier' : template_identifier}
    return post('registerTemplate', params, port)

def new_template(template_identifier, data, port):
    params = {'templateIdentifier' : template_identifier, 'templateData' : data}
    return post('newTemplate', params, port)

def generate(document_identifier, template_identifier, fragment_identifier, data, port):
    params = {'documentIdentifier' : document_identifier,
            'templateIdentifier' : template_identifier,
            'fragmentIdentifier' : fragment_identifier, 'data' : data}
    return post('generate', params, port)

def list_documents(port):
    return get('list', {}, port)

def delete(identifier, port):
    params = {'identifier' : identifier}
    return post('delete', params, port)

def export(identifier, doctype, port):
    params = {'identifier' : identifier, 'docType' : doctype}
    return get('export', params, port)

def geninit(document_identifier, template_identifier, fragment_identifier, data, port):
    params = {'documentIdentifier' : document_identifier,
            'templateIdentifier' : template_identifier,
            'fragmentIdentifier' : fragment_identifier, 'data' : data}
    return post('geninit', params, port)

def genzone(document_identifier, template_identifier, fragment_identifier, zone_identifier, data, port):
    params = {'documentIdentifier' : document_identifier,
            'templateIdentifier' : template_identifier,
            'fragmentIdentifier' : fragment_identifier,
            'zoneIdentifier' : zone_identifier, 'data' : data}
    return post('genzone', params, port)

def genfinal(document_identifier, template_identifier, fragment_identifier, port):
    params = {'documentIdentifier' : document_identifier,
            'templateIdentifier' : template_identifier,
            'fragmentIdentifier' : fragment_identifier}
    return post('genfinal', params, port)

def simulate(template_identifier, limits, port):
    params = {'templateIdentifier' : template_identifier, 'limits' : limits}
    return post('simulate', params, port)

def test1():
    dn = 'test-data1'
    new(dn, 'http://name.space')
    namespace(dn, 'sub', 'http://sub.name.space')

    f = open('examples/multizone.json')
    t = f.read()
    f.close()
    register(dn, 'test-temp1', t)

    f = open('examples/multizone-subst.json')
    s = f.read()
    f.close()
    generate(dn, 'test-temp1', 'test-frag1', s)

    f = open('examples/multizone-subst-0.json')
    s = f.read()
    f.close()
    geninit(dn, 'test-temp1', 'test-frag2', s)

    for z, n in [(z, n) for z in ['sz', 'pz'] for n in range(1, 3)]:
        f = open('examples/multizone-subst-{}-{}.json'.format(z, n))
        s = f.read()
        f.close()
        genzone(dn, 'test-temp1', 'test-frag2', ':' + z, s)

    genfinal(dn, 'test-temp1', 'test-frag2')

    for ident in json.loads(list_documents()):
        print ident
    print

    d = export(dn, 'document')
    print d

    delete('test-temp1')
    delete(dn)

def test2():
    dn = 'test-data1'
    new(dn, 'http://name.space')
    namespace(dn, 'sub', 'http://sub.name.space')

    f = open('examples/multizone.json')
    t = f.read()
    f.close()
    register(dn, 'test-temp1', t)

    s = simulate('test-temp1', json.dumps({':sz' : 5, ':pz' : 5}))
    print s

    generate(dn, 'test-temp1', 'test-frag1', s)

    d = export(dn, 'document')
    print d

    delete('test-temp1')
    delete(dn)

if __name__ == '__main__':
    test1()
    test2()
